### PR TITLE
Enable sampling from multidimensional logits

### PR DESCRIPTION
**Pull request by ishihara-y**
Improve softmax distribution and support sampling from multidimensional logits.

### CODE REVIEW

**Review by sbsekiguchi — APPROVED** [review]
LGTM.
